storage/innobase/include/trx0sys.h

```diff
@@ -418,6 +420,7 @@ class rw_trx_hash_t
             (trx_state_eq(trx, TRX_STATE_ACTIVE) &&
              (!srv_was_started ||
               srv_read_only_mode ||
+              !trx_rollback_is_active ||
```

Comment:
It seems that we fail to invoke `rollback_all_recovered_task.wait()` in `logs_empty_and_mark_files_at_shutdown()` or `srv_shutdown_threads()`. It seems that this added condition might introduce a resource leak if the rollback is still running during shutdown. I would suggest to replace the following code in `logs_empty_and_mark_files_at_shutdown()`:
```c++
	/* We need these threads to stop early in shutdown. */
	const char* thread_name = srv_fast_shutdown != 2
		&& trx_rollback_is_active
		? "rollback of recovered transactions" : nullptr;
```
with something like this (note: I hope that it will not hang if there was no prior call to `srv_thread_pool->submit_task(&rollback_all_recovered_task)` in case startup was aborted due to some error):
```c++
	if (srv_fast_shutdown != 2) {
		rollback_all_recovered_task.wait();
		ut_ad(!trx_rollback_is_active);
	}
```
This logic will have to be tested with various values of `innodb_force_recovery` and `innodb_fast_shutdown`.

I would also suggest to minimize the references to `trx_rollback_is_active` in other code. Ideally we would only have a few `ut_ad(!trx_rollback_is_active)` where appropriate.